src/protocols/secure_channel/CASESession.cpp

```diff
@@ -404,187 +445,283 @@ CHIP_ERROR CASESession::HandleSigmaR1(System::PacketBufferHandle && msg)
+                                                static_cast<uint32_t>(msg_r2_signed_enc_len + kTAGSize)));
+    ReturnErrorOnFailure(tlvWriterMsg2.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriterMsg2.Finalize(&msg_R2));
 
-    err = mCommissioningHash.AddData(ByteSpan{ msg_R2->Start(), msg_R2->DataLength() });
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(mCommissioningHash.AddData(ByteSpan{ msg_R2->Start(), msg_R2->DataLength() }));
 
-    mNextExpectedMsg = Protocols::SecureChannel::MsgType::CASE_SigmaR3;
+    mState = kSentSigma2;
 
     // Call delegate to send the msg to peer
-    err = mExchangeCtxt->SendMessage(Protocols::SecureChannel::MsgType::CASE_SigmaR2, std::move(msg_R2),
-                                     SendFlags(SendMessageFlags::kExpectResponse));
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(mExchangeCtxt->SendMessage(Protocols::SecureChannel::MsgType::CASE_Sigma2, std::move(msg_R2),
+                                                    SendFlags(SendMessageFlags::kExpectResponse)));
 
-    ChipLogDetail(SecureChannel, "Sent SigmaR2 msg");
+    ChipLogDetail(SecureChannel, "Sent Sigma2 msg");
 
-exit:
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CASESession::HandleSigma2Resume(System::PacketBufferHandle && msg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    uint16_t responderSessionId;
+
+    uint32_t decodeTagIdSeq = 0;
+
+    ChipLogDetail(SecureChannel, "Received Sigma2Resume msg");
+
+    uint8_t sigma2ResumeMIC[kTAGSize];
+
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    SuccessOrExit(err = tlvReader.GetBytes(mResumptionId, kCASEResumptionIDSize));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    SuccessOrExit(err = tlvReader.GetBytes(sigma2ResumeMIC, kTAGSize));
+
+    SuccessOrExit(err = ValidateSigmaResumeMIC(ByteSpan(sigma2ResumeMIC), ByteSpan(mInitiatorRandom), ByteSpan(mResumptionId),
+                                               ByteSpan(kKDFS2RKeyInfo), ByteSpan(kResume2MIC_Nonce)));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    SuccessOrExit(err = tlvReader.Get(responderSessionId));
+
+    ChipLogDetail(SecureChannel, "Peer assigned session key ID %d", responderSessionId);
+    SetPeerSessionId(responderSessionId);
+
+    SendStatusReport(mExchangeCtxt, kProtocolCodeSuccess);
```

Comment:
Suggest adding a TODO to handle the following spec text:

> The initiator SHALL set <<ref_SecureSessionContext, `SessionTimestamp`>> to a timestamp from a clock which would allow for the eventual determination of the last session use relative to other sessions.` 

This applies to Sigma3 and and SigmaFinished processing as well, on the correct side. This is to eventually allow selecting a least-recently-used session for eviction on running out of contexts.